returnn/forward.py

```diff
@@ -207,3 +216,188 @@ def hash(cls, kwargs):
+        self.returnn_root = returnn_root
+        self.log_verbosity = log_verbosity
+        self.device = device
+
+        self.out_returnn_config_file = self.output_path("returnn.config")
+        self.out_files = {output: self.output_path(output) for output in output_files}
+
+        self.rqmt = {
+            "gpu": 1 if device == "gpu" else 0,
+            "cpu": cpu_rqmt,
+            "mem": mem_rqmt,
+            "time": time_rqmt,
+        }
+
+    def tasks(self):
+        yield Task("create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        """create files"""
+        config = self.create_returnn_config(
+            model_checkpoint=self.model_checkpoint,
+            returnn_config=self.returnn_config,
+            log_verbosity=self.log_verbosity,
+            device=self.device,
+        )
+        config.write(self.out_returnn_config_file.get_path())
+
+        cmd = [
+            self.returnn_python_exe.get_path(),
+            os.path.join(self.returnn_root.get_path(), "rnn.py"),
+            self.out_returnn_config_file.get_path(),
+        ]
+        util.create_executable("rnn.sh", cmd)
+
+        # check here if model actually exists
+        if self.model_checkpoint is not None:
+            assert os.path.exists(
+                _get_model_path(self.model_checkpoint).get_path()
+            ), f"Provided model checkpoint does not exists: {self.model_checkpoint}"
+
+    def run(self):
+        """run"""
+        # run everything in a TempDir as writing files can cause heavy load
+        with tempfile.TemporaryDirectory(prefix=gs.TMP_PREFIX) as tmp_dir:
+            print("using temp-dir: %s" % tmp_dir)
+            call = [
+                self.returnn_python_exe.get_path(),
+                os.path.join(self.returnn_root.get_path(), "rnn.py"),
+                self.out_returnn_config_file.get_path(),
+            ]
+
+            try:
+                env = os.environ.copy()
+                env["OMP_NUM_THREADS"] = str(self.rqmt["cpu"])
+                env["MKL_NUM_THREADS"] = str(self.rqmt["cpu"])
+                subprocess.check_call(call, cwd=tmp_dir, env=env)
+            except Exception:
+                print("Run crashed - copy temporary work folder as 'crash_dir'")
+                if os.path.exists("crash_dir"):
+                    shutil.rmtree("crash_dir")
+                shutil.copytree(tmp_dir, "crash_dir")
+                raise
+
+            # move logs
+            if os.path.exists(f"{tmp_dir}/returnn.log"):
+                shutil.move(f"{tmp_dir}/returnn.log", "returnn.log")
+            if os.path.exists(f"{tmp_dir}/returnn-tf-log"):
+                shutil.move(f"{tmp_dir}/returnn-tf-log", ".")
+
+            # move outputs to output folder
+            for k, v in self.out_files.items():
+                assert os.path.exists(f"{tmp_dir}/{k}"), f"Output file {k} does not exist"
```

Comment:
I'm not exactly sure I understand your question. You are asking about the `assert` one line before, how this can fail? This fails e.g. if the user specifies `outfiles = ["my_data.txt"]` or so for the job, and then RETURNN does not create this file because you had some bug in the config.

No, we cannot check beforehand, without having a full replication of all logic of RETURNN and running the config, basically running the job.

This can also execute users code. Which is even the standard case, where the user has custom callbacks in the config for the forwarding which would write to files. It's as hard as the halting problem.